feat: add oauth2 token exchange

code-asher opened the pull request:
This adds a new identity provider package and hooks it into coderd routes along with some test refactoring.  See commit messages for more details.

There is more work to be done (tracking in https://github.com/coder/coder/pull/11609) but I think this is fully featured enough to be merged in.  All the routes are only accessible in dev mode.

It is getting pretty big, although 60% is HTML, tests, and generated code!

Blocked by:
- https://github.com/coder/coder/pull/12196

Blocks:
- https://github.com/coder/coder/pull/11780
- https://github.com/coder/coder/pull/12127

code-asher committed:
Extract auth code helper

code-asher committed:
Add static OAuth2 authorize page

code-asher committed:
Add URL query validator

code-asher committed:
Allow multiple required query params

code-asher committed:
Refactor test apps

They are named now so their purpose is clearer.

Also, getting apps can be done as a user so do that instead.

code-asher committed:
Add OAuth2 auth, token, and revoke routes

code-asher committed:
Add OAuth2 app filtering by user

code-asher committed:
make gen

code-asher committed:
Allow fetching app with query param and form value

code-asher committed:
Check that required query params are non-zero

code-asher commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,245 @@
+func Hash(secret string, id uuid.UUID) []byte {
+	b := []byte(secret)
+	// TODO: Expose iterations, memory, and threads as configuration values?
+	return argon2.IDKey(b, []byte(id.String()), 1, 64*1024, 2, uint32(len(b)))
```
I am using the app ID for this but I think the salt is supposed to be unique per secret.  I could prepend the ID of every secret to the secret, although that makes them pretty long.  Only the client secret is ever seen by a person though.

Emyrk replied:
I think we need to ditch this and use `userpassword.Hash`. I think it might be fine in practice, I just rather stick to the hash method we already know and use. I left another comment about how to handle the fact that `userpassword.Hash` is non-deterministic. 

Some notes on the security:
- `userpassword.Hash` is stronger because users might have the same password (`password123`) and these should be hashed to different values.
- This has function takes in a `secret` which already has a lot of randomness and is longer, so it does not need as strong of a hash function.

Still would rather use what we already use.

Emyrk replied:
> I am using the app ID for this but I think the salt is supposed to be unique per secret. 

For the best security yes. You could use `sha256(appID + userID)[:10]` or something deterministic but more unique. The best is a random salt, like done in `userpassword.Hash()`

> I could prepend the ID of every secret to the secret, although that makes them pretty long

We can make a new ID that is shorter. I put this comment above, but it just came to me if you want to use the `uuid` already and make it smaller, just do: `sha256(id)[:10]` and encode it with like base64 or something. The downside to just making a new ID for this purpose those is that base64 has is a non deterministic length encoding for any given input. (well you can predict it, but it is a pain).

So if we just choose random characters, then the ID is always a fixed length. We should not just use a `uuid` string though, as it is a terribly inefficient use of characters and is way to long.

Emyrk commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,246 @@
+
+func authorizationCodeGrant(ctx context.Context, db database.Store, app database.OAuth2ProviderApp, defaultLifetime time.Duration, clientSecret, code string) (oauth2.Token, error) {
+	// Validate the client secret.
+	secretHash := Hash(clientSecret, app.ID)
+	secret, err := db.GetOAuth2ProviderAppSecretByAppIDAndSecret(
+		//nolint:gocritic // Users cannot read secrets so we must use the system.
+		dbauthz.AsSystemRestricted(ctx),
+		database.GetOAuth2ProviderAppSecretByAppIDAndSecretParams{
+			AppID:        app.ID,
+			HashedSecret: secretHash[:],
+		})
```
Ooooh I now see why you did this. My above comment about using `userpassword.Hash` remains, I think we should just prefix this stuff with some id and use `userpassword.Compare(hashedSecret, secret)` to validate.

This would mean you would have to store an extra id in the database...

So currently a secret is 40 random characters. We should prefix it with something.

Before: `EA9ILdVv0wVgXo840dFZV4UX21dtZsl5rziDiiU7`

New: `coder_io928hVamL_EA9ILdVv0wVgXo840dFZV4UX21dtZsl5rziDiiU7` (or something like this).

You would have to store `io928hVamL` in the db, so when we are given the secret, we split on `_`, find the `id` and do the search with that. This column also needs a `unique` constraint.

Emyrk commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	// Validate the authorization code.
+	codeHash := Hash(code, app.ID)
```
Same as above hash comment

Emyrk commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,246 @@
+		return oauth2.Token{}, sql.ErrNoRows
+	}
+
+	// Generate a refresh token.
+	// The refresh token is not currently used or exposed though as API keys can
+	// already be refreshed by just using them.
+	// TODO: However, should we implement the refresh grant anyway?
+	// 40 characters matches the length of GitHub's client secrets.
+	rawRefreshToken, err := cryptorand.String(40)
+	if err != nil {
+		return oauth2.Token{}, err
+	}
```
Maybe we disable auto refresh for oauth type api keys :shrug:.

Emyrk commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,246 @@
+			return xerrors.Errorf("insert oauth2 access token: %w", err)
+		}
+
+		hashed := Hash(rawRefreshToken, app.ID)
```
Same hash comment as above

Emyrk commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,246 @@
+	return oauth2.Token{
+		AccessToken: sessionToken,
+		TokenType:   "Bearer",
+		// TODO: Exclude until refresh grant is implemented.
+		// RefreshToken: rawRefreshToken,
+		// Expiry:       key.ExpiresAt,
```
I see why you do not have Expiry, but that usually means it never expires.

I think we should add a expiry and disable auto refreshing of these api keys.

code-asher replied:
Oh yeah I was trying to see if omitting it would make Backstage stop trying to refresh, but it tries to anyway.  I wanted to merge this and work on refresh next but maybe I have to add it now.

Emyrk commented on enterprise/coderd/oauth2.go:
```diff
@@ -45,15 +46,47 @@ func (api *API) oAuth2ProviderMiddleware(next http.Handler) http.Handler {
 		return
 	}
+
+	var dbApps []database.OAuth2ProviderApp
+	for _, app := range userApps {
+		dbApps = append(dbApps, database.OAuth2ProviderApp{
+			ID:          app.OAuth2ProviderApp.ID,
+			Name:        app.OAuth2ProviderApp.Name,
+			CallbackURL: app.OAuth2ProviderApp.CallbackURL,
+			Icon:        app.OAuth2ProviderApp.Icon,
+		})
+	}
 	httpapi.Write(ctx, rw, http.StatusOK, db2sdk.OAuth2ProviderApps(api.AccessURL, dbApps))
```
We might update the db app, and we should pass the whole db app to the db2sdk func always. This is more resistance to migrations on the BE.


```suggestion
	var sdkApps []codersdk.OAuth2ProviderApp
	for _, app := range userApps {
		sdkApps = append(sdkApps, db2sdk.OAuth2ProviderApp(api.AccessURL, app.OAuth2ProviderApp))
	}
	httpapi.Write(ctx, rw, http.StatusOK, sdkApps)
```

code-asher replied:
Ahhh makes sense!  Much better.

Emyrk commented on enterprise/coderd/oauth2.go:
```diff
@@ -208,7 +241,7 @@ func (api *API) postOAuth2ProviderAppSecret(rw http.ResponseWriter, r *http.Requ
 		return
 	}
-	hashed := sha256.Sum256([]byte(rawSecret))
+	hashed := identityprovider.Hash(rawSecret, app.ID)
```
hash comments above

Emyrk reviewed (changes requested):
Consolidating and adding some hash function notes.

We should use `userpasword.Hash()`

The user secret is currently 40 random characters, let's spruce that up.
- Add a `coder_` prefix so a random key in a text file is more easily recognized. 
- Add an `id` section of 10(?) random characters with a unique constraint.
- Optionally... if you want to be really cool, add a 2 character checksum at the end, so if someone hand types this we can detect typos :wink:

code-asher commented on coderd/httpapi/queryparams.go:
```diff
@@ -233,7 +246,7 @@ func ParseCustomList[T any](parser *QueryParamParser, vals url.Values, def []T,
 	parser.addParsed(queryParam)
 	// If the query param is required and not present, return an error.
-	if parser.RequiredParams[queryParam] && (!vals.Has(queryParam)) {
+	if parser.RequiredParams[queryParam] && (!vals.Has(queryParam) || vals.Get(queryParam) == "") {
```
Yeah I had mixed feelings about this; it felt surprising to me that I could write `Required("client_id")` and it could come through empty.  My thinking was that if I want to require a query param I also want a value.

I did check existing usages; it is used for `reconnect` on the terminal (which is invalid if blank) and for `start_time` and `end_time` on an insights endpoint which also cannot be blank (it will fail to be parsed into a time).

But at the same time I see the distinction between required and empty.  An empty query value is probably only useful as a boolean where the presence of the key means `true`, and I could see requiring a `true` for something like `?acceptTerms` or something like that before you can use the endpoint.

So let me rename this to `RequiredNotEmpty`, that will leave room for `Required` if we do need that some day.

code-asher replied:
Some new ones popped up after I merged main: `template_id`, `workspace_id`, and `agent_id` on some endpoints, they all make database queries with these values so I think they would just return "no rows" or a 404 or whatever, so this change will make it return "invalid query params, template_id is required" instead which I think is fair, but it does change the semantics a little from a 404 to a 400.

code-asher committed:
Delete some oidc helper comments

code-asher committed:
s/Required/RequiredNotEmpty

code-asher commented on codersdk/oauth2.go:
```diff
@@ -168,3 +179,50 @@ func (c *Client) DeleteOAuth2ProviderAppSecret(ctx context.Context, appID uuid.U
+}
+
+// RevokeOAuth2ProviderApp completely revokes an app's access for the user.
+func (c *Client) RevokeOAuth2ProviderApp(ctx context.Context, appID uuid.UUID) error {
+	res, err := c.Request(ctx, http.MethodDelete, "/login/oauth2/tokens", nil, func(r *http.Request) {
+		q := r.URL.Query()
+		q.Set("client_id", appID.String())
+		r.URL.RawQuery = q.Encode()
+	})
```
Great point, I will add a note to implement that.

Yup, it is 1 token per user per app.

code-asher committed:
No URL for me

code-asher commented on enterprise/coderd/identityprovider/authorize.go:
```diff
@@ -0,0 +1,160 @@
+	// We add "redirected" when coming from the authorize page.
+	_ = p.String(vals, "", "redirected")
+
+	if err := validateRedirectURL(cb, params.redirectURL.String()); err != nil {
+		p.Errors = append(p.Errors, codersdk.ValidationError{
+			Field:  "redirect_uri",
+			Detail: fmt.Sprintf("Query param %q is invalid", "redirect_uri"),
+		})
+	}
```
Ooo good point, let me move this.

code-asher committed:
Move redirect check into query parser

It will now be checked on the token side as well, although we do not
actually use it there anyway.  But this seems more right.

The URL parser was added for this OAuth2 code so we can just rename it.

code-asher commented on enterprise/coderd/identityprovider/authorize.go:
```diff
@@ -0,0 +1,160 @@
+		})
+	}
+
+	p.ErrorExcessParams(vals)
+	return params, p.Errors, nil
```
Makes sense!  That does seem better.

code-asher replied:
Oh right I needed to check the length from the caller to tell if it was the `url.Parse(callback)` that failed (although that should not happen because we validate before storing it in the db) or if it was a validation error so I can give a different message and error code.  I think I will just lift the parse out of this func.

Emyrk replied:
You can also return a sentinel error. So either a unique type or a unique message.

code-asher replied:
Ohh right that is smart!

code-asher committed:
Return err on invalid params

That way the caller can just check err instead of the length of errors.

code-asher commented on enterprise/coderd/identityprovider/authorize.go:
```diff
@@ -0,0 +1,160 @@
+	return params, p.Errors, nil
+}
+
+/**
+ * Authorize displays an HTML page for authorizing an application when the user
+ * has first been redirected to this path and generates a code and redirects to
+ * the app's callback URL after the user clicks "allow" on that page, which is
+ * detected via the origin and referer headers.
+ */
```
Gah, thanks, I keep mixing up my doc styles.

code-asher committed:
Use correct godoc style

code-asher commented on enterprise/coderd/identityprovider/authorize.go:
```diff
@@ -0,0 +1,160 @@
+			})
+			return
+		}
+		hashedCode := Hash(rawCode, app.ID)
```
This is actually a relief lol

code-asher committed:
Use userpassword.Hash

code-asher commented on enterprise/coderd/identityprovider/authorize.go:
```diff
@@ -0,0 +1,160 @@
+		}
+		hashedCode := Hash(rawCode, app.ID)
+		err = db.InTx(func(tx database.Store) error {
+			// Delete any previous codes.
+			err = tx.DeleteOAuth2ProviderAppCodesByAppAndUserID(ctx, database.DeleteOAuth2ProviderAppCodesByAppAndUserIDParams{
+				AppID:  app.ID,
+				UserID: apiKey.UserID,
+			})
+			if err != nil && !errors.Is(err, sql.ErrNoRows) {
+				return xerrors.Errorf("delete oauth2 app codes: %w", err)
+			}
```
Oh yup you are absolutely right.  I think I got confused thinking I had to delete to cascade something but the app codes are all on their own, unlike tokens.  Will change!

code-asher commented on enterprise/coderd/identityprovider/authorize.go:
```diff
@@ -0,0 +1,160 @@
+			_, err = tx.InsertOAuth2ProviderAppCode(ctx, database.InsertOAuth2ProviderAppCodeParams{
+				ID:        uuid.New(),
+				CreatedAt: dbtime.Now(),
+				// TODO: Configurable expiration?  Ten minutes matches GitHub.
+				ExpiresAt:    dbtime.Now().Add(time.Duration(10) * time.Minute),
```
Ah yup good point, will do.

code-asher committed:
Add comments to code timeout

code-asher committed:
Comment on blank origin

code-asher commented on enterprise/coderd/identityprovider/middleware.go:
```diff
@@ -0,0 +1,117 @@
+func authorizeMW(accessURL *url.URL) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			origin := r.Header.Get(httpmw.OriginHeader)
+			originU, err := url.Parse(origin)
+			if err != nil {
+				// TODO: Curl requests will not have this. One idea is to always show
+				// html here??
+				httpapi.Write(r.Context(), rw, http.StatusBadRequest, codersdk.Response{
+					Message: "Invalid or missing origin header.",
+					Detail:  err.Error(),
+				})
+				return
+			}
```
Oh wow good to know.

code-asher commented on enterprise/coderd/identityprovider/middleware.go:
```diff
@@ -0,0 +1,117 @@
+				//    better to validate now to avoid wasting the user's time clicking a
+				//    button that will just error anyway.
+				params, errs, err := extractAuthorizeParams(r, app.CallbackURL)
+				if err != nil {
+					site.RenderStaticErrorPage(rw, r, site.ErrorPageData{
+						Status:       http.StatusInternalServerError,
+						HideStatus:   false,
+						Title:        "Internal Server Error",
+						Description:  err.Error(),
+						RetryEnabled: false,
+						DashboardURL: accessURL.String(),
+						Warnings:     nil,
+					})
+					return
+				}
+				if len(errs) > 0 {
+					errStr := make([]string, len(errs))
+					for i, err := range errs {
+						errStr[i] = err.Detail
+					}
+					site.RenderStaticErrorPage(rw, r, site.ErrorPageData{
+						Status:       http.StatusBadRequest,
+						HideStatus:   false,
+						Title:        "Invalid Query Parameters",
+						Description:  "One or more query parameters are missing or invalid.",
+						RetryEnabled: false,
+						DashboardURL: accessURL.String(),
+						Warnings:     errStr,
+					})
+					return
+				}
```
I refactored so we are always checking `err`.  I do still have separate error since one is if the callback URL fails to parse, which should not happen but is technically a different error if it does.

code-asher commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,246 @@
+
+		var token oauth2.Token
+		//nolint:gocritic,revive // More cases will be added later.
+		switch params.grantType {
+		// TODO: Client creds, device code, refresh.
+		default:
+			token, err = authorizationCodeGrant(ctx, db, app, defaultLifetime, params.clientSecret, params.code)
```
Oh yeah, I did have it this way but I am validating the grant type in the parser now so it is not possible to get anything other than a valid type or empty string.

code-asher commented:
Moved to a draft while I work on:

1. [ ] Changing the hashing
2. [ ] Adding the refresh flow

code-asher committed:
Pass the whole app to db2sdk

code-asher committed:
Fix a racy context

And rename another, it is not being used incorrectly, but the name makes
it clear.

code-asher committed:
Extract state from authURL

code-asher committed:
Implement refresh grant

code-asher committed:
Fix verbiage in app PUT

code-asher committed:
Remove extra period

code-asher committed:
Fix test race

code-asher commented:
Rebased, but only commits from [Delete some oidc helper comments](https://github.com/coder/coder/pull/11778/commits/cbce34ac532022488576610854888dfc6c881a6a) onward are new.  They should all be self-contained enough to review individually, if that appeals.

I really like the idea of a checksum but I have not implemented it yet.  I was reading up some check digit algorithms and thinking maybe of using crc8 or something but got hit with decision paralysis so I am leaving it for the moment.  :sweat:

code-asher committed:
Apparently browsers do not always set origin

The referer check was not quite right either.

Emyrk reviewed (approved):
This is really good :+1:. 

Just 1 thing to address before I approve it in `refreshTokenGrant`. Is there a leak of refresh codes?

Emyrk reviewed (changes requested):
whoops meant to do request changes

code-asher commented on coderd/httpapi/queryparams.go:
```diff
@@ -121,6 +123,27 @@ func (p *QueryParamParser) UUIDs(vals url.Values, def []uuid.UUID, queryParam st
+	if v.Host != base.Host || !strings.HasPrefix(v.Path, base.Path) {
+		p.Errors = append(p.Errors, codersdk.ValidationError{
+			Field:  queryParam,
+			Detail: fmt.Sprintf("Query param %q is invalid", queryParam),
```
Good idea!

code-asher committed:
Mention redirect URL must be a subset

code-asher committed:
Remove redundant comment

code-asher committed:
Clarify revoke operates on an authorized user

code-asher commented on codersdk/oauth2.go:
```diff
@@ -168,3 +179,50 @@ func (c *Client) DeleteOAuth2ProviderAppSecret(ctx context.Context, appID uuid.U
+
+// RevokeOAuth2ProviderApp completely revokes an app's access for the user.
+func (c *Client) RevokeOAuth2ProviderApp(ctx context.Context, appID uuid.UUID) error {
+	res, err := c.Request(ctx, http.MethodDelete, "/login/oauth2/tokens", nil, func(r *http.Request) {
```
Oh really good point, let me change it now real quick.

code-asher replied:
Oh right I had it here to have the post/delete paradigm on the same endpoint.  Maybe I should actually move this back to `/api/v2` since it is our own thing, not part of the whole oauth flow.

Emyrk replied:
This is not an oauth2 spec endpoint right? 

Eh just keep it here and comment that it is not oauth 2 spec. If you feel this has to be a static endpoint though, then just comment why it should be. I'll defer to your judgement

code-asher committed:
Move cURL comment

code-asher commented on enterprise/coderd/identityprovider/middleware.go:
```diff
@@ -0,0 +1,129 @@
+			app := httpmw.OAuth2ProviderApp(r)
+			ua := httpmw.UserAuthorization(r)
+
+			// url.Parse() allows empty URLs, which is fine because the origin is not
+			// always set by browsers (or other tools like cURL).  If the origin does
+			// exist, we will make sure it matches.  We require `referer` to be set at
+			// a minimum, however.
+			cameFromSelf := (origin == "" || originU.Hostname() == accessURL.Hostname()) &&
+				refererU.Hostname() == accessURL.Hostname() &&
+				refererU.Path == "/login/oauth2/authorize"
```
Oh yeah that does sound like a nightmare.  Hmmm...so `referer` is only required to detect the `allow` part of it, and it could be blank on the initial HTML render if opened from a desktop app or something I think.  So maybe I error if both redirected is set and referer is blank, since the redirected param indicates the user might have pressed the button.

There is a todo for cURL at the top of the function!  But I should move it down next to the HTML part.

Emyrk replied:
> So maybe I error if both redirected is set and referer is blank, since the redirected param indicates the user might have pressed the button.

This is perfect. I was thinking the exact same thing on why it would be difficult. But this solution is perfect for now imo.

We had a customer that bitbucket had this error because their proxy was nuking the Referrer header. It was the hardest thing in the world to debug.

code-asher committed:
Add error when referer is blank

code-asher commented on enterprise/coderd/identityprovider/revoke.go:
```diff
@@ -0,0 +1,44 @@
+			httpapi.InternalServerError(rw, err)
+			return
+		}
+		httpapi.Write(ctx, rw, http.StatusNoContent, nil)
```
I think it was because there is no body.

code-asher committed:
Comment app secret struct

code-asher committed:
Use var block

code-asher commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,370 @@
+	refreshToken string
+}
+
+func extractTokenParams(r *http.Request, callbackURL *url.URL) (tokenParams, []codersdk.ValidationError, error) {
+	p := httpapi.NewQueryParamParser()
+	err := r.ParseForm()
+	if err != nil {
+		return tokenParams{}, nil, xerrors.Errorf("parse form: %w", err)
+	}
+
+	vals := r.Form
+	p.RequiredNotEmpty("grant_type")
+	grantType := httpapi.ParseCustom(p, vals, "", "grant_type", httpapi.ParseEnum[codersdk.OAuth2ProviderGrantType])
+	switch grantType {
+	case codersdk.OAuth2ProviderGrantTypeRefreshToken:
+		p.RequiredNotEmpty("refresh_token")
+	case codersdk.OAuth2ProviderGrantTypeAuthorizationCode:
+		p.RequiredNotEmpty("client_secret", "client_id", "code")
+	}
+
+	params := tokenParams{
+		clientID:     p.String(vals, "", "client_id"),
+		clientSecret: p.String(vals, "", "client_secret"),
+		code:         p.String(vals, "", "code"),
+		grantType:    grantType,
+		redirectURL:  p.RedirectURL(vals, callbackURL, "redirect_uri"),
+		refreshToken: p.String(vals, "", "refresh_token"),
+	}
+
+	p.ErrorExcessParams(vals)
+	if len(p.Errors) > 0 {
+		return tokenParams{}, p.Errors, xerrors.Errorf("invalid query params: %w", p.Errors)
+	}
+	return params, nil, nil
+}
```
It really is!!  :heart:

code-asher commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,370 @@
+	vals := r.Form
+	p.RequiredNotEmpty("grant_type")
+	grantType := httpapi.ParseCustom(p, vals, "", "grant_type", httpapi.ParseEnum[codersdk.OAuth2ProviderGrantType])
+	switch grantType {
+	case codersdk.OAuth2ProviderGrantTypeRefreshToken:
+		p.RequiredNotEmpty("refresh_token")
+	case codersdk.OAuth2ProviderGrantTypeAuthorizationCode:
+		p.RequiredNotEmpty("client_secret", "client_id", "code")
+	}
```
Yeah you are right, let me add that.

code-asher committed:
Add error for unhandled grant types

Emyrk commented on enterprise/coderd/identityprovider/tokens.go:
```diff
@@ -0,0 +1,370 @@
+
+	// Replace the token.
+	err = db.InTx(func(tx database.Store) error {
+		ctx := dbauthz.As(ctx, userSubj)
+		err = tx.DeleteAPIKeyByID(ctx, prevKey.ID) // This cascades to the token.
+		if err != nil {
+			return xerrors.Errorf("delete oauth2 app token: %w", err)
+		}
+
+		newKey, err := tx.InsertAPIKey(ctx, key)
+		if err != nil {
+			return xerrors.Errorf("insert oauth2 access token: %w", err)
+		}
+
+		_, err = tx.InsertOAuth2ProviderAppToken(ctx, database.InsertOAuth2ProviderAppTokenParams{
+			ID:          uuid.New(),
+			CreatedAt:   dbtime.Now(),
+			ExpiresAt:   key.ExpiresAt,
+			HashPrefix:  []byte(refreshToken.Prefix),
+			RefreshHash: []byte(refreshToken.Hashed),
+			AppSecretID: dbToken.AppSecretID,
+			APIKeyID:    newKey.ID,
+		})
+		if err != nil {
+			return xerrors.Errorf("insert oauth2 refresh token: %w", err)
+		}
+		return nil
```
Ah! If it cascades, we are good.
